File: src/motion_py_pkg/motion_py_pkg/manipulator/create_rot_z.py
import numpy as np


def create_rot_z(th, thoff):
    """
    Create the 4x4 homogeneous transformation matrix for a rotation around the Z axis.

    Parameters:
        th    (float): Angle of rotation in radians.
        thoff (float): Angle offset in radians (handles pi/2 and -pi/2 special cases).

    Returns:
        np.ndarray: 4x4 transformation matrix.
    """
    if np.isclose(thoff, -np.pi / 2):
        rotation = np.array([
            [ np.sin(th),  np.cos(th), 0],
            [-np.cos(th),  np.sin(th), 0],
            [          0,           0, 1],
        ])
    elif np.isclose(thoff, np.pi / 2):
        rotation = np.array([
            [-np.sin(th), -np.cos(th), 0],
            [ np.cos(th), -np.sin(th), 0],
            [          0,           0, 1],
        ])
    else:
        rotation = np.array([
            [np.cos(th), -np.sin(th), 0],
            [np.sin(th),  np.cos(th), 0],
            [         0,           0, 1],
        ])

    translation = np.array([[0], [0], [0]])
    homogeneous = np.array([[0, 0, 0, 1]])

    return np.block([[rotation, translation], [homogeneous]])

File: src/manipulator_matlab/createRotZ.m
function RotZ = createRotZ(th, thoff)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% function RotZ = createRotZ(th)
% 
% Task: create the 3D transformation matrix corresponding to a rotation around the z axis
%
% Inputs:
%	- th: the angle of rotation
%   - thoff: the offset of rotation
%
% Output: 
%	- RotZ: the transformation matrix corresponding to a rotation around the z axis
%
% author: Sébastien Dubois, s.j.dubois.116@cranfield.ac.uk
% date: 12/12/2025
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Create the rotation part of the transformation matrix (3 x 3)
if thoff==-pi/2
    rotationMatrix = [sin(th) cos(th) 0;
        -cos(th) sin(th) 0
        0 0 1];
elseif thoff==pi/2
    rotationMatrix = [-sin(th) -cos(th) 0;
        cos(th) -sin(th) 0
        0 0 1];
else
    rotationMatrix = [cos(th) -sin(th) 0;
        sin(th) cos(th) 0
        0 0 1];
end

%% create the translation part (3 x 1)
translationMatrix = [0; 0; 0];

%% create the homogeneous coordinate part
homogeneousCoord = [0 0 0 1];

%% Express the final tranformation matrix for a rotation around z
RotZ = [rotationMatrix translationMatrix; homogeneousCoord];